src/lib/transaction_snark/test/account_timing/account_timing.ml

```diff
@@ -297,4 +296,576 @@ let%test_module "account timing check" =
+         run on multiple keypairs
+      *)
+      let length = 1 in
+      (* keypair pair (for payment sender/receiver)  *)
+      let keypairss =
+        List.init length ~f:(fun _ ->
+            (Signature_lib.Keypair.create (), Signature_lib.Keypair.create ()))
+      in
+      (* list of keypairs *)
+      let keypairs =
+        List.map keypairss ~f:(fun (kp1, kp2) -> [ kp1; kp2 ]) |> List.concat
+      in
+      (keypairss, keypairs)
+
+    (* we're testing timings, not signatures *)
+    let validate_user_command uc =
+      let (`If_this_is_used_it_should_have_a_comment_justifying_it validated_uc)
+          =
+        Signed_command.to_valid_unsafe uc
+      in
+      validated_uc
+
+    (* Mina_ledger.Ledger.copy does not actually copy *)
+    let copy_ledger (ledger : Mina_ledger.Ledger.t) =
+      let ledger_copy =
+        Mina_ledger.Ledger.create ~depth:(Mina_ledger.Ledger.depth ledger) ()
+      in
+      let accounts = Mina_ledger.Ledger.to_list ledger in
+      List.iter accounts ~f:(fun account ->
+          let pk = Account.public_key account in
+          let token = Account.token account in
+          let account_id = Account_id.create pk token in
+          match
+            Mina_ledger.Ledger.get_or_create_account ledger_copy account_id
+              account
+          with
+          | Ok (`Added, _loc) ->
+              ()
+          | Ok (`Existed, _loc) ->
+              failwithf
+                "When creating ledger, account with public key %s and token %s \
+                 already existed"
+                (Signature_lib.Public_key.Compressed.to_string pk)
+                (Token_id.to_string token) ()
+          | Error err ->
+              failwithf
+                "When creating ledger, error adding account with public key %s \
+                 and token %s: %s"
+                (Signature_lib.Public_key.Compressed.to_string pk)
+                (Token_id.to_string token) (Error.to_string_hum err) ()) ;
+      ledger_copy
+
+    let check_transaction_snark ~(txn_global_slot : Mina_numbers.Global_slot.t)
+        (ledger : Mina_ledger.Ledger.t)
+        (transaction : Mina_transaction.Transaction.t) =
+      let sok_message =
+        Sok_message.create ~fee:Currency.Fee.zero
+          ~prover:
+            Public_key.(compress (of_private_key_exn (Private_key.create ())))
+      in
+      let precomputed_values = Lazy.force Precomputed_values.compiled_inputs in
+      let state_body0 =
+        Mina_state.(
+          With_hash.data precomputed_values.protocol_state_with_hashes
+          |> Protocol_state.body)
+      in
+      let consensus_state0 =
+        Mina_state.Protocol_state.Body.consensus_state state_body0
+      in
+      let txn_state_view0 = Mina_state.Protocol_state.Body.view state_body0 in
+      (* based on Sparse_ledger.handler
+         don't need a reference to the ledger, it's mutated
+      *)
+      let handler ledger =
```

Comment:
Use `Sparse_ledger.handler` and `Sparse_ledger.of_ledger_subset_exn`? similar to what's done here https://github.com/MinaProtocol/mina/blob/fcc92f0f561a625c277fbfa21bcbe72b2ee6f28f/src/lib/transaction_snark/test/transaction_union/transaction_union.ml#L128